src/QueryAccess/QueryAccessHandlerBase.php

```diff
@@ -0,0 +1,243 @@
+
+    $conditions = NULL;
+    if ($operation == 'view' && $this->entityType->entityClassImplements(EntityPublishedInterface::class)) {
+      $published_key = $this->entityType->getKey('published');
+      if ($entity_conditions) {
+        // Restrict the existing conditions to published entities only.
+        $conditions = new ConditionGroup('AND');
+        $conditions->addCacheContexts(['user.permissions']);
+        $conditions->addCondition($entity_conditions);
+        $conditions->addCondition($published_key, '1');
+      }
+    }
+    else {
+      $conditions = $entity_conditions;
+    }
+
+    if (!$conditions) {
+      // The user doesn't have access to any entities.
+      // Falsify the query to ensure no results are returned.
+      $conditions = new ConditionGroup('OR');
+      $conditions->addCacheContexts(['user.permissions']);
+      $conditions->alwaysFalse();
+    }
+
+    return $conditions;
+  }
+
+  /**
+   * Builds the conditions for entities that have an owner.
+   *
+   * @param string $operation
+   *   The access operation. One of "view", "update" or "delete".
+   * @param \Drupal\Core\Session\AccountInterface $account
+   *   The user for which to restrict access.
+   *
+   * @return \Drupal\entity\QueryAccess\ConditionGroup|null
+   *   The conditions, or NULL if the user doesn't have access to any entity.
+   */
+  protected function buildEntityOwnerConditions($operation, AccountInterface $account) {
+    $entity_type_id = $this->entityType->id();
+    $uid_key = $this->entityType->getKey('uid');
+    $bundle_key = $this->entityType->getKey('bundle');
+
+    $conditions = new ConditionGroup('OR');
+    $conditions->addCacheContexts(['user.permissions']);
+    // Any $entity_type permission.
+    if ($account->hasPermission("$operation any {$entity_type_id}")) {
+      // The user has full access, no conditions needed.
+      return $conditions;
+    }
+
+    $bundles = array_keys($this->bundleInfo->getBundleInfo($entity_type_id));
+    $bundles_with_any_permission = [];
+    foreach ($bundles as $bundle) {
+      if ($account->hasPermission("$operation any $bundle $entity_type_id")) {
+        $bundles_with_any_permission[] = $bundle;
+      }
+    }
+    // Any $bundle permission.
+    if ($bundles_with_any_permission) {
+      $conditions->addCondition($bundle_key, $bundles_with_any_permission);
+    }
+
+    // Own $entity_type permission.
+    if ($account->hasPermission("$operation own $entity_type_id")) {
+      $conditions->addCacheContexts(['user']);
+      $conditions->addCondition($uid_key, $account->id());
+    }
+
+    // Own $bundle permission.
+    foreach ($bundles as $bundle) {
+      if ($account->hasPermission("$operation own $bundle $entity_type_id")) {
+        $conditions->addCacheContexts(['user']);
+        $conditions->addCondition((new ConditionGroup('AND'))
+          ->addCondition($uid_key, $account->id())
+          ->addCondition($bundle_key, $bundle)
+        );
+      }
+    }
```

Comment:
In case we're not re-using the code from buildEntityConditions(), the above loop could be pulled into the other bundle loop, populate a  $bundles_with_own_permission array and then add that as follows:
```    // Own $bundle permission.
    if ($bundles_with_own_permission) {
        $conditions->addCacheContexts(['user']);
        $conditions->addCondition((new ConditionGroup('AND'))
          ->addCondition($uid_key, $account->id())
          ->addCondition($bundle_key, $bundles_with_own_permission)
        );
    }
```

Right now, you're adding the same condition group for every bundle.

Reply:
Good point.